src/lib/ldap/base.c

```diff
@@ -884,6 +930,50 @@ int fr_ldap_global_config(int debug_level, char const *tls_random_file)
 	return 0;
 }
 
+/** Free any libldap structures when an fr_ldap_query_t is freed
+ *
+ */
+static int _ldap_query_free(fr_ldap_query_t *query)
+{
+	/*
+	 *	If any referrals were followed, the parsed referral URLS should be freed
+	 */
+	if (query->referral_urls) ldap_memvfree((void **)query->referral_urls);
+
+	if (query->result) ldap_msgfree(query->result);
+
+	if (query->ldap_url) {
+		/*
+		 *	Queries originating from parsed URLs may have parsed
+		 *	server controls as well
+		 */
+		if ((query->ldap_url->lud_exts) && (query->serverctrls[0])) ldap_controls_free(query->serverctrls);
```

Comment:
We probably always want to free the client and server controls, not just when query->ldap_url->lud_exts is set?

Reply:
Sometimes the server controls are set from static data and sometimes parsed from URLs - maybe the solution is a flag for each to indicate they need freeing ..?

Reply:
There's an equivalent function `fr_ldap_control_clear(fr_ldap_connection_t *conn)` for connections.